JBPM-5962 Stunner - custom bendpoints on connectors

tiagodolphine opened the pull request:
Now the control points (bendpoints) are being handled by stunner based on Commands, so it is possible to Add, Move, Remove them with all the undo, redo actions. 
The marshalling was updated to support the control points, so it is possible to save/open diagram keeping the control points.
Besides the control points feature, some code improvement that involves many refactors were made, so it will be necessary to make regression tests do guarantee none other features were broken, please.
Depends on https://github.com/kiegroup/lienzo-core/pull/2 to be merged.

@romartin 
@hasys

manstis commented:
@tiagodolphine can you please confirm DMN's ```DirectionalLine``` cannot have bend points inserted? DMN connectors should only have head/tail and the User should not be able to add bends. Thanks.

tiagodolphine commented:
Hey @manstis  yeah I confirm it, the bendpoints are disabled  on the DMN due to a restriction I set on the `DirectionalLine.isControlPointShape` to false. This is checked when applying the bendpoints on the connector. 
Thanks.

romartin commented:
Hey @tiagodolphine , please synchronize with @Josephblt as well - he's doing the (regular) polyline connector implementation, he's working on Lienzo side, and next we'll need the stunner polyline class and bendpoints handling for it as well.

romartin commented:
hey @hasys , I'm just doing review for this... added you as well for review, that's an important feature added and it also contains important bug fixes, so it must be well tested... you think should I add any other reviewer? Thx!

romartin commented:
Hey @tiagodolphine 

Did a try to the WAR your provided. I found a few bugs in the first try... so I think it needs a review from your side again, I don't think those are difficult bugs, just need to do some last cleans you know. Anyway that's a great job, thanks for the fixings it contains as well, really nice :+1: 

Added a video in the following zip file:
[bendpoints-issues.zip](https://github.com/kiegroup/kie-wb-common/files/1697266/bendpoints-issues.zip)

As you can see in the video:
- I'm missing the CP for the default point that connectors create right now. I mean if you realize, when creating a connector it uses like  3 or 4 CPs by default, so the user get the ortogonal lines working out-of-the-box. Just creating a new task from the event and moving it around you see that the connector has different parts, that "intersection" points are not managed by the user, it does not appear the red circle... does that makes sense?
- After updating some CP and undoing the action, it's not possible to re-do it ,clicking on the button does not complain (probably the command exist on the stack), but does nothing
- After removing a connector and undoing - it gets broken
- Finally, which I'm not really sure about and if makes sense can be considered optional, the connector's CPs are not being part or the parent shape right? I mean if I drag a connector inside a parent and drag the parent, the CPs keep on same location. 

Thanks!!

romartin commented:
@tiagodolphine  - Another improvement that comes into my mind... listening for arrow keys for moving control points as well as we do with shapes... makes sense? sounds feasible?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/UpdateControlPointPositionCommand.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    protected AbstractCompositeCommand<GraphCommandExecutionContext, RuleViolation> initialize(GraphCommandExecutionContext context) {
+        addCommand(new DeleteControlPointCommand(candidate, controlPoint));
```
mmm so updating the CP location does not updates its index, right? Why not just updating the location for it instead of adding and removing it?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/shape/view/wires/ext/WiresShapeViewExt.java:
```diff
@@ -222,6 +226,24 @@ public T showControlPoints(final ControlPointType type) {
 
+    @Override
+    public List<ControlPoint> addControlPoint(ControlPoint... controlPoint) {
+        //TODO: check if this is applied here
```
should we implement this?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/shape/view/wires/ext/WiresShapeViewExt.java:
```diff
@@ -222,6 +226,24 @@ public T showControlPoints(final ControlPointType type) {
+
+    @Override
+    public T removeControlPoint(ControlPoint... controlPoint) {
+        //TODO: check if this is applied here
```
should we implement this?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/shape/view/wires/ext/WiresShapeViewExt.java:
```diff
@@ -222,6 +226,24 @@ public T showControlPoints(final ControlPointType type) {
+
+    @Override
+    public List<ControlPoint> getShapeControlPoints() {
+        //TODO: check if this is applied here
```
should we implement this?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/wires/StunnerWiresConnectorHandler.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public void onNodeMouseDoubleClick(NodeMouseDoubleClickEvent nodeMouseDoubleClickEvent) {
+        //skip the double click handle here, this will be handled on the {#link org.kie.workbench.common.stunner.core.client.canvas.controls.connection.ControlPointControl}
```
the idea of other handlers is about firing events which finally the "CPControl" observers, why not same with the double click event? which I would say that is used for both adding/removing a CP, right?

tiagodolphine replied:
Yes we can do in this way, I just thought that it will be easier to customize to other action instead of double click, if we think it is feasible on the future, but it is easy to handle this on lienzo as well, but then it is tied to double click only. WDYT ?

tiagodolphine commented:
Jenkins retest this

tiagodolphine commented:
Hey @romartin follow are the updates on your comments:

_I'm missing the CP for the default point that connectors create right now. I mean if you realize, when creating a connector it uses like 3 or 4 CPs by default, so the user get the ortogonal lines working out-of-the-box. Just creating a new task from the event and moving it around you see that the connector has different parts, that "intersection" points are not managed by the user, it does not appear the red circle... does that makes sense?_
***Current we don't have any default control point even before this PR, we can add that, but maybe opening a new JIRA for that, WDYT?**

_After updating some CP and undoing the action, it's not possible to re-do it ,clicking on the button does not complain (probably the command exist on the stack), but does nothing_
**Solved.**

_After removing a connector and undoing - it gets broken_
**Solved**

_Finally, which I'm not really sure about and if makes sense can be considered optional, the connector's CPs are not being part or the parent shape right? I mean if I drag a connector inside a parent and drag the parent, the CPs keep on same location._
***I tried to simply resolve on Stunner side with commands but it didn't work, because on the UpdateElementPositionCommand we just move the parent and the children are handled on Lienzo, so maybe the best solution is to handle the Control Points on Lienzo as well. Maybe a new JIRA because it will demand some time on this.**

_What about listening for arrow keys for moving control points as well as we do with shapes... makes sense? sounds feasible?_
** Yes, this can be implemeted but we need to insert a handler for onMouseClick on the ControlPoint (on liezo, same way as drag), may be a new JIRA WDYT ?**

_About Viewer/Editor modes_
 **I just updated the MapSelectionControl to check is the binded session was not ClientFullSession, then the update the shape state to HIGHLIGHT insted of SELECTED (this bug happened even withou the control points using the connection magnets). The other control points actions are not enabled because the ClientReadOnlySessionImpl does not inject the ControlPointControl.**

_WiresShapeViewExt_
**Solved, created a new interface to manage control points.**

Thanks.

tiagodolphine commented:
Jenkins execute full downstream build

tiagodolphine commented:
Jenkins retest this

tiagodolphine commented:
Jenkins execute full downstream build

romartin commented:
Hey @tiagodolphine  I would update the PR's description, as it still refers to the manually generated WAR and to the `com.ahome-it` PR's ... I mean just for not creating confusion to others. Up to you - you can remove that comments and just say it depends on the new lienzo-core and lienzo-tests PRs, so adding PR links.. This way I expect people testing using the generated WARs on Jenkins too. WDYT?

tiagodolphine commented:
Hey @romartin yes, I did remove the comment that it depends on the com.ahome-it PR's.
The tests now can be made using the generated WARs on the CI.
Thanks.

romartin commented:
Hey @tiagodolphine 

Yep agree on your comments, let's open new tickets for identified improvements Also just did another to try this stuff, and now it works much better! :+1:  

Just a few comments:

* **Found bug** -. it's not possible to update a connection. For example, changing the target connection to a certain task magnet results in an exception,:
![screenshot from 2018-02-14 00-44-12](https://user-images.githubusercontent.com/4602417/36180525-28414cb0-1121-11e8-81d3-d9183a68d135.png)

* **Performance** - Not sure since when/which commit/why yet but once you open a stunner diagram, do whatever stuff, then close and reopen - the editor because really slow! I was wondering if you fixes would solve that issue, but it's still there.. does it happens to you as well? any ideas? did you analyzed it? Otherwise np, let's open a BLOCKER ticket just for this, it's not due to this commit anyway

* `Instance Disposers` - looking at code I see you're now injecting and passing around `Disposer` instances to fix some memory issues, that's fine! we're not doing that before.... :/ On the other hand for client side stuff we use to inject the beans by using `ManagedInstance` from Errai, which it already provides "disposer" methods for a single instance or for all  instances produced... so just wondering if it's really necessary to pass the Disposers around to the controllers or just pass the `ManagedInstance`'s. Anyway not really sure about this, I don't have the detail.. just minor comment to think about hehe :)

So rather than the bug found, I think it's all fine for me.

Please @hasys , do a good try to this, it contains the new feature about bendpoints but also a few fixes in other parts, so better if you can double check and do some smoke testing for it - so waiting for you review too! 

Thanks guys!! :+1:

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/EdgeShape.java:
```diff
@@ -16,9 +16,14 @@
 import org.kie.workbench.common.stunner.core.graph.Edge;
 import org.kie.workbench.common.stunner.core.graph.Node;
+import org.kie.workbench.common.stunner.core.graph.content.view.ControlPoint;
+import org.kie.workbench.common.stunner.core.graph.content.view.Point2D;
```
Useless imports additions.

tiagodolphine replied:
Done.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/marshall/json/builder/AbstractEdgeBuilder.java:
```diff
@@ -116,6 +127,22 @@ protected void afterEdgeBuild(final BuilderContext context,
+                    Counter indexCounter = new Counter(0);
+                    ControlPoint[] controlPoints = dockers.subList(1, dockers.size() - 1).stream()
+                            .sequential()
+                            .map(docker -> Objects.equals(docker.length, 2) ? new Point2D(docker[0], docker[1]) : null)
```
It is pretty much the same, but I changed to use == to simplify since we are comparing int,int

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/marshall/json/builder/AbstractEdgeBuilder.java:
```diff
@@ -116,6 +127,22 @@ protected void afterEdgeBuild(final BuilderContext context,
                 }
+
+                //ControlPoints
+                if(dockers.size()>2) {
```
Ok, refactored extracting into methods.

tiagodolphine commented:
@romartin I fixed the issue you found when changing the connection points.
About perfomance I agree, we really need to solve that, let's work on that, on the leak JIRA.
And the Disposer, yeah, this is one way to destroy CDI beans, we can use the ManagedInstance as well, the result will be the same it is a matter of choice, if we decide to do that I can change.

@hasys about your comments the first two are related and in fact they are related to the issue @romartin found as well, and to solve that we need changes on Lienzo, that will consider the connector and control points when moving a parent shape, I gave the suggestion to open a  JIRA for that since this PR is already to long and complex, WDYT ?
And regarding the third issue, I fixed now there are no exception, but the delete behavior on multiselection is not that good in my point of view, since it is deleting only nodes, not the connectors related to the nodes, but this is other issue note related to bendpoints, so we can open a JIRA for that WDYT @romartin @hasys ?

Thank you guys for the comments and the review !

tiagodolphine commented:
Jenkins execute full downstream build

romartin commented:
Yes please @tiagodolphine , @hasys  I would vote for identifying and opening new issues on JIRA, if are not related to this task, and try to get this PR merged asap. This commit is big and @Josephblt also depends on it, so it's important getting this one merged...

LuboTerifaj commented:
Hey, I am testing it, will let you know.

LuboTerifaj commented:
It seems that full-downstream build failed. Can you fix it please?

tiagodolphine commented:
@LuboTerifaj it seems the error is not related and war files for the kie-wb were generated, anyway I 'll trigger the build again, let's see.

tiagodolphine commented:
Jenkins execute full downstream build

LuboTerifaj commented:
It looks good to me. Just one small observation. Imho, bended points don't work as good as they could. When click on point I must wait for a moment. If don't wait, it is impossible to move with it. Simply you can not take point and drag it in one moment. I am not sure if this is bug or it is intended. Shift-click also doesn't work as it is described in Jira [1]. Double-click works.
[1] https://issues.jboss.org/browse/JBPM-5962

LuboTerifaj commented:
@tiagodolphine I realized that later. Thank you.

tiagodolphine commented:
Hey @LuboTerifaj  thanks for the review. 
Yeah, it is set to double-click, this is the default behavior on lienzo, but we can change that on Stunner side, IMHO double-click seems more intuitive, @romartin WDYT? 
About the waiting time to drag it, I assume that is on lienzo, maybe we can open other improvement JIRA for this, I think this is not a blocker for now, WDYT @LuboTerifaj , @romartin ?  to avoid delaying to much this PR.

LuboTerifaj commented:
@tiagodolphine yes, sure it is not blocker. I just wanted to mention it.

romartin commented:
@LuboTerifaj @hasys waiting for your approvals to merge this, when you think it's ready.. thanks!

LuboTerifaj commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/shape/impl/ConnectorShape.java:
```diff
@@ -69,11 +73,32 @@ public void applyConnections(final Edge<ViewConnector<W>, Node> element,
+    }
+
+    private HasManageableControlPoints getShapeViewWithControlPoints() {
+        if(!(getShapeView() instanceof HasManageableControlPoints)){
```
Formatting: Missing spaces would be nice here.

tiagodolphine replied:
Done

LuboTerifaj commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/LienzoMultipleSelectionControl.java:
```diff
@@ -168,7 +168,9 @@ protected void onDisable() {
 
     private void clear() {
-        getSelectionManager().clearSelection();
+        if(Objects.nonNull(getSelectionControl().getCanvasHandler())) {
```
Formatting: Missing space after "if" would be nice here.

tiagodolphine replied:
Done

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/graph/content/HasControlPoints.java:
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Done

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/AbstractCanvasGraphCommand.java:
```diff
@@ -123,26 +141,45 @@
+
+        final Command<GraphCommandExecutionContext, RuleViolation> graphCommand = getGraphCommand(context);
+        CommandResult<RuleViolation> graphResult = null;
+        switch (op) {
```
@LuboTerifaj this switch-case is based on the enum ` CommandOperation` that has only these 3 values, so in this case there will be no default block, we should consider all the case possibilities.

LuboTerifaj replied:
OK. Thank you for clarification.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/AbstractCanvasGraphCommand.java:
```diff
@@ -123,26 +141,45 @@
+        getGraphCommand(context);
+
+        final Command<AbstractCanvasHandler, CanvasViolation> command = getCanvasCommand(context);
+        switch (op) {
```
the same situation here, all the cases on the enum should be handled.

LuboTerifaj replied:
OK. Thank you for clarification.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/AddCanvasConnectorCommand.java:
```diff
@@ -40,15 +43,20 @@ public AddCanvasConnectorCommand(final Edge candidate,
+        context.register(shapeSetId, candidate);
+        context.applyElementMutation(candidate, MutationContext.STATIC);
+
+        if(candidate.getContent() instanceof ViewConnector){
```
Done

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/AddCanvasConnectorCommand.java:
```diff
@@ -40,15 +43,20 @@ public AddCanvasConnectorCommand(final Edge candidate,
+            ControlPoint[] controlPoints = ((ViewConnector) candidate.getContent()).getControlPoints().stream().toArray(ControlPoint[]::new);
+            CommandResult<CanvasViolation> addControlPointsResult = new AddCanvasControlPointCommand(candidate, controlPoints).execute(context);
+            ShapeUtils.hideControlPoints(candidate, context);
+            if(CommandUtils.isError(addControlPointsResult)){
```
Done

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/UpdateCanvasControlPointPositionCommand.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Override
+    protected AbstractCompositeCommand<AbstractCanvasHandler, CanvasViolation> initialize(AbstractCanvasHandler context) {
+        if(ShapeUtils.getControlPoints(edge, context).stream().anyMatch(cp-> Objects.equals(cp.getLocation(), position))){
```
Done

tiagodolphine commented:
@LuboTerifaj just updated the PR with a new commit.
Thanks for the review.

romartin commented:
Jenkins execute full downstream build
